README


This archive contains the distribution Protocol-SOCKS,
version 0.002:

  abstract SOCKS protocol support

This software is copyright (c) 2014 by Tom Molesworth.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for Protocol-SOCKS

0.002     2014-08-23 19:51:37+01:00 Europe/London
	No new features.

	Test fixes:
	* Missing tests from previous release

0.001     2014-08-23 19:40:37+01:00 Europe/London
	Initial CPAN release


lib/Protocol/SOCKS/Client.pm
package Protocol::SOCKS::Client;
$Protocol::SOCKS::Client::VERSION = '0.002';
use strict;
use warnings;

use parent qw(Protocol::SOCKS);

=head1 NAME

Protocol::SOCKS::Client - client support for SOCKS protocol

=head1 VERSION

Version 0.002

=head1 DESCRIPTION

This provides an abstraction for dealing with the client side of the SOCKS protocol.

=cut

use Future;
use Socket qw(inet_pton inet_ntop inet_ntoa AF_INET AF_INET6);

use Protocol::SOCKS::Constants qw(:all);

=head1 METHODS

=cut

=head2 completion

Returns the completion future.

=cut

sub completion { $_[0]->{completion} ||= $_[0]->new_future }

=head2 auth

Returns the auth Future.

=cut

sub auth { $_[0]->{auth} ||= $_[0]->new_future }

=head2 auth_methods

Returns the list of auth methods we can handle.

=cut

sub auth_methods {
	my $self = shift;
	@{ $self->{auth_methods} ||= [ AUTH_NONE ] }
}

=head2 init_packet

Initial client packet.

=cut

sub init_packet {
	my $self = shift;
	my @methods = (0);
	pack 'C1C/C*', $self->version, $self->auth_methods;
}

=head2 on_read

Handler for reading data from the server.

=cut

sub on_read {
	my ($self, $buf) = @_;
	if(!$self->auth->is_ready) {
		return unless length($$buf) >= 2;
		my ($version, $method) = unpack 'C1C1', substr $$buf, 0, 2, '';
		die "Unexpected version" unless $version == $self->version;
		if($method == 0xFF) {
			$self->auth->fail($method);
		} else {
			$self->auth->done($method);
		}
		return;
	} else {
		# warn "non-auth, have " . length($$buf) . "bytes";
		return unless my ($host, $port) = $self->parse_reply($buf);

		my $f = shift @{$self->{awaiting_reply}};
		$f->done($host, $port);
	}
}

=head2 init

Startup - writes the initial packet to the server.

=cut

sub init {
	my $self = shift;
	$self->write($self->init_packet);
}

=head2 connect

Issues a connection request.

=cut

sub connect {
	my ($self, $atype, $addr, $port) = @_;
	my $f = $self->new_future;
	my $opaque_addr = $self->pack_address($atype, $addr);
	push @{$self->{awaiting_reply}}, $f;
	$self->write(
		pack(
			'C1C1C1',
			$self->version,
			0x01,
			0x00,
		) . $opaque_addr . pack('n1', $port)
	);
	$f;
}

=head2 parse_reply

Parse a server reply.

=cut

sub parse_reply {
	my ($self, $buffref) = @_;
	return unless length $$buffref >= 4;
	my ($version, $status, $reserved, $atype) = unpack 'C1C1C1C1', substr $$buffref, 0, 4;
	if($status != 0) {
		# warn $Protocol::SOCKS::REPLY_CODE{$status};
		return;
	}

	substr $$buffref, 0, 3, '';
	my $addr = $self->extract_address($buffref);
	my $port = unpack 'n1', substr $$buffref, 0, 2, '';
	# warn "Addr $addr, port $port\n";
	return $addr, $port;
}

1;

__END__

=head1 AUTHOR

Tom Molesworth <cpan@perlsite.co.uk>

=head1 LICENSE

Copyright Tom Molesworth 2014. Licensed under the same terms as Perl itself.


lib/Protocol/SOCKS/Server.pm
package Protocol::SOCKS::Server;
$Protocol::SOCKS::Server::VERSION = '0.002';
use strict;
use warnings;

use parent qw(Protocol::SOCKS);

=head1 NAME

Protocol::SOCKS::Server - server support for SOCKS protocol

=head1 VERSION

Version 0.002

=head1 DESCRIPTION

This provides an abstraction for dealing with the server side of the SOCKS protocol.

=cut

use Future;
use Socket qw(inet_pton inet_ntop inet_ntoa AF_INET AF_INET6);

use Protocol::SOCKS::Constants qw(:all);

=head1 METHODS

=cut

=head2 completion

Returns the completion future.

=cut

sub completion { $_[0]->{completion} ||= $_[0]->new_future }

=head2 auth

Returns the auth Future.

=cut

sub auth { $_[0]->{auth} ||= $_[0]->new_future }

=head2 auth_methods

Returns the list of auth methods we can handle.

=cut

sub auth_methods {
	my $self = shift;
	@{ $self->{auth_methods} ||= [ AUTH_NONE ] }
}

=head2 init_packet

Initial client packet.

=cut

sub init_packet {
	my $self = shift;
	my @methods = (0);
	pack 'C1C/C*', $self->version, $self->auth_methods;
}

=head2 on_read

Handler for reading data from the client.

=cut

sub on_read {
	my ($self, $buf) = @_;
	if(!$self->init->is_ready) {
		return if length($$buf) < 3;
		my (undef, $method_count) = unpack 'C1C', substr $$buf, 0, 2;
		return unless length($$buf) >= 2 + $method_count;

		my ($version, $methods) = unpack 'C1C/C*', substr $$buf, 0, 2 + $method_count, '';
		die "Invalid version" unless $version == $self->version;
		my $auth_method;
		METHOD:
		for my $method (split //, $methods) {
			next METHOD unless grep $method == $_, $self->auth_methods;
			$auth_method = $method;
			last METHOD;
		}
		unless(defined $auth_method) {
			$self->write(
				pack 'C1C1',
					$self->version,
					AUTH_FAIL,
			);
			return $self->init->fail(auth => 'no suitable methods');
		}
		$self->init->done($version => $auth_method);
		return $self->write(
			pack 'C1C1',
				$self->version,
				$auth_method
		)
	}

	return unless my $details = $self->parse_request($buf);

	my $f = shift @{$self->{awaiting_reply}};
	$f->done($details);
}

=head2 init

Resolves with version and auth method when connection
has been established

=cut

sub init { $_[0]->{init} ||= $_[0]->new_future }

=head2 parse_request

Parse a client request.

=cut

sub parse_request {
	my ($self, $buffref) = @_;
	return unless length $$buffref >= 6;
	my ($version, $cmd, $reserved, $atype) = unpack 'C1C1C1C1', substr $$buffref, 0, 4;
	die "unknown command $cmd" unless $cmd > 0 && $cmd < 4;

	substr $$buffref, 0, 3, '';
	my $addr = $self->extract_address($buffref);
	my $port = unpack 'n1', substr $$buffref, 0, 2, '';
	warn "Addr $addr, port $port\n";
}

1;

__END__

=head1 AUTHOR

Tom Molesworth <cpan@perlsite.co.uk>

=head1 LICENSE

Copyright Tom Molesworth 2014. Licensed under the same terms as Perl itself.


lib/Protocol/SOCKS.pm
package Protocol::SOCKS;
# ABSTRACT: abstract SOCKS protocol support
use strict;
use warnings;

our $VERSION = '0.002';

=head1 NAME

Protocol::SOCKS - abstract support for the SOCKS5 network protocol

=head1 VERSION

Version 0.002

=head1 DESCRIPTION

=cut

use Future;
use Socket qw(inet_pton inet_ntop inet_ntoa AF_INET AF_INET6);

use Protocol::SOCKS::Constants qw(:all);

our %REPLY_CODE = (
	0x00 => 'succeeded',
	0x01 => 'general SOCKS server failure',
	0x02 => 'connection not allowed by ruleset',
	0x03 => 'Network unreachable',
	0x04 => 'Host unreachable',
	0x05 => 'Connection refused',
	0x06 => 'TTL expired',
	0x07 => 'Command not supported',
	0x08 => 'Address type not supported',
);

=head1 METHODS

=cut

=head2 new

Instantiates this protocol object.

=cut

sub new { my $class = shift; bless { @_ }, $class }

=head2 version

Our protocol version. Usually 5.

=cut

sub version { shift->{version} ||= 5 }

=head2 write

Called when we want to write data. Requires a writer to be configured.

=cut

sub write { $_[0]->{writer}->($_[1]) }

=head2 new_future

Instantiates a new L<Future> using the provided factory, or calls through to L<Future>->new.

=cut

sub new_future { (shift->{future_factory} ||= sub { Future->new })->() }

=head2 pack_fqdn

Packs a fully-qualified domain into a data structure.

=cut

sub pack_fqdn {
	my $self = shift;
	$self->pack_address(ATYPE_FQDN, @_)
}

=head2 pack_ipv4

Packs an IPv4 address into a data structure.

=cut

sub pack_ipv4 {
	my $self = shift;
	$self->pack_address(ATYPE_IPV4, @_)
}

=head2 pack_ipv6

Packs an IPv6 address into a data structure.

=cut

sub pack_ipv6 {
	my $self = shift;
	$self->pack_address(ATYPE_IPV6, @_)
}

=head2 pack_address

Packs an address of the given type into a data structure.

=cut

sub pack_address {
	my ($self, $type, $addr) = @_;
	if($type == ATYPE_IPV4) {
		return pack('C1', $type) . inet_pton(AF_INET, $addr);
	} elsif($type == ATYPE_IPV6) {
		return pack('C1', $type) . inet_pton(AF_INET6, $addr);
	} elsif($type == ATYPE_FQDN) {
		return pack('C1C/a*', $type, $addr);
	} else {
		die sprintf 'unknown address type 0x%02x', $type;
	}
}

=head2 extract_address

Extracts address information from a scalar ref.

=cut

sub extract_address {
	my ($self, $buf) = @_;
	return undef unless length($$buf) > 1;

	my ($type) = unpack 'C1', substr $$buf, 0, 1;
	if($type == ATYPE_IPV4) {
		return undef unless length($$buf) >= (1 + 4);
		(undef, my $ip) = unpack 'C1N1', substr $$buf, 0, 1 + 4, '';
		return '' unless $ip;
		warn "== $ip\n";
		return inet_ntoa($ip);
	} elsif($type == ATYPE_IPV6) {
		return undef unless length($$buf) >= (1 + 16);
		(undef, my $ip) = unpack 'C1A16', substr $$buf, 0, 1 + 16, '';
		return inet_ntop(AF_INET6, $ip);
	} elsif($type == ATYPE_FQDN) {
		my ($len) = unpack 'C1', substr $$buf, 1, 1;
		return undef unless length($$buf) >= (1 + 1 + $len);
		(undef, my $host) = unpack 'C1C/a*', substr $$buf, 0, 1 + 1 + $len, '';
		return $host;
	} else {
		die sprintf 'unknown address type 0x%02x', $type;
	}
}

1;

__END__

=head1 SEE ALSO

=over 4

=item * L<Net::SOCKS>

=item * L<POE::Component::Client::SOCKS>

=item * L<POE::Component::Proxy::SOCKS>

=item * L<LWP::Protocol::socks>

=item * L<IO::Stream::Proxy::SOCKSv5>

=item * L<IO::Socket::Socks>

=back

=head1 AUTHOR

Tom Molesworth <cpan@perlsite.co.uk>

=head1 LICENSE

Copyright Tom Molesworth 2014. Licensed under the same terms as Perl itself.
